[receiver/journald] add support for running journalctl in a chroot

jefferbrecht opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
Adds two new config options to the journald receiver that allow the collector, when in container mode, to run `journalctl` in a chroot mounted from the host:

* `root_path`: path to a mounted chroot environment from the host, e.g. configure this as `/host` when running the collector under docker with `-v /:/host`
* `journalctl_path`: path to the journalctl binary within the chroot environment, e.g. `/usr/bin/journalctl`

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Resolves #43731

<!--Describe what testing was performed and which tests were added.-->
#### Testing
* Added unit tests for the new configuration options
* Built a collector image locally and verified that the new config options work as expected by mounting the host's rootfs and calling its journalctl from the collector to collect real journald logs

<!--Describe the documentation added.-->
#### Documentation
Documentation for the feature is captured in the receiver's README.

<!--Please delete paragraphs that you did not use before submitting.-->

jefferbrecht committed:
[receiver/journald] add root_path config option

This allows users to run journalctl from a chroot.

Running from a chroot is a robust way to work around the
problems associated with bundling a journalctl binary into a
collector container. Namely, this allows a containerized collector
to use a journalctl binary mounted from the host instead, without
needing to individually mount all associated shared libraries and
host configurations.

jefferbrecht committed:
[receiver/journald] add journalctl_path config option

This allows users to specify an exact path to the
journalctl binary.

In particular, this is needed when using the
root_path config option due to a [Golang issue](https://github.com/golang/go/issues/39341).

belimawr reviewed (changes requested):
@jefferbrecht, that's a great contribution, thanks!

Because of https://github.com/golang/go/issues/39341, if `root_path` is set and `journalctl_path` is *not* set, the collector cannot start journalctl.

I believe it would be good to add a configuration validation to ensure that if `root_path` is set, then `journalctl_path` is an absolute path. What do you think?

Something like the following added to `pkg/stanza/operator/input/journald/config_linux.go`:
```go
func (cfg *Config) Validate() error {
	if cfg.RootPath != "" {
		if !filepath.IsAbs(cfg.JournalctlPath) {
			return errors.New("'journalctl_path' must be an absolute path when 'root_path' is set")
		}
	}

	return nil
}

```

jefferbrecht commented:
> @jefferbrecht, that's a great contribution, thanks!
> 
> Because of [golang/go#39341](https://github.com/golang/go/issues/39341), if `root_path` is set and `journalctl_path` is _not_ set, the collector cannot start journalctl.
> 
> I believe it would be good to add a configuration validation to ensure that if `root_path` is set, then `journalctl_path` is an absolute path. What do you think?
> 
> Something like the following added to `pkg/stanza/operator/input/journald/config_linux.go`:
> 
> ```go
> func (cfg *Config) Validate() error {
> 	if cfg.RootPath != "" {
> 		if !filepath.IsAbs(cfg.JournalctlPath) {
> 			return errors.New("'journalctl_path' must be an absolute path when 'root_path' is set")
> 		}
> 	}
> 
> 	return nil
> }
> ```

Sure, I'm on board with adding a validation for that. Most of the config validation today is done in `buildArgs()` (or `buildNewCmdFunc()` as I've wrapped it in this PR) though; do you think it makes more sense to put your suggested validation in there and then call `buildNewCmdFunc()` from `Validate()`?

namco1992 commented:
> No we should not call it ourselves, Validate() is already called by the Collector before instantiating the receiver, so defining it is enough to cause the Collector to exit on error, we don't need to call it in the input itself.

The journald receiver consists of two parts: a thin wrapper in the `receiver/journaldreceiver` to comply with the receiver pattern, and a stanza input in `pkg/stanza/operator/input/journald` where all the heavy lifting is done. **This also means that we have 2 "configs" here, one in the receiver, and one in the stanza input.**

For most receivers, the config implements the `Validator` interface in https://github.com/open-telemetry/opentelemetry-collector/blob/main/confmap/xconfmap/config.go, and the collector calls the `Validate` method recursively to validate the config. I suppose that's what @belimawr means here.

However, it seems the receivers that rely on stanza inputs went on a different path. The receiver config itself doesn't implement the `Validator` interface, nor does any stanza inputs' config.

We can, however, implement the `Validator` interface in the journald receiver, but it seems to be an outlier among other stanza-dependent receivers. E.g. for filelog receiver, the receiver config looks like this (and no `Validate()` implemented):
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/5523e29729b3596eb58d348157d9939cb1ce435e/receiver/filelogreceiver/filelog.go#L52-L58

And the validation happens during the `Build`:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/5523e29729b3596eb58d348157d9939cb1ce435e/pkg/stanza/fileconsumer/config.go#L101-L104

tl;dr I think we should do the validation inside the `Build` func, align with other stanza-dependent receivers. It could be a private `validate()` method like what filelog receiver does, or just dump it in the `Build` (e.g. https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/operator/input/tcp/config.go), or `buildArgs()` as @jefferbrecht suggested. I don't really have a strong opinion on this.

In terms of exploring other patterns for config validation, IMO it could be a separate topic. FWIW, I do feel the current stanza-dependent receiver pattern is a bit convoluted.

jefferbrecht committed:
[receiver/journald] move validation to validate()

This helps separate validation from command-building. I'm leaving
the buildMatchesConfig() validation out for now because it happens
quite deep inside that function and isn't cleanly separable yet.

The readme is also clarified with restrictions on root_path and
journalctl_path.

jefferbrecht commented:
Thanks all -- I've added a `c.validate()` and moved (most of) the validations there. I left `buildMatchesConfig()`'s validation out for now because it happens very deep and isn't cleanly separable yet.

belimawr reviewed (commented):
It looks great! I only have a small comment about an error message.

jefferbrecht committed:
[receiver/journald] generalize root_path error message

jefferbrecht committed:
Merge branch 'main' into jefferbrecht-journald-chroot

jefferbrecht commented on pkg/stanza/operator/input/journald/config_linux.go:
```diff
@@ -26,31 +31,58 @@ func init() {
+		info, err := os.Stat(c.RootPath)
+		if err != nil {
+			return fmt.Errorf("root_path %q not found: %w", c.RootPath, err)
+		}
```
Done!

braydonk committed:
Merge branch 'main' into jefferbrecht-journald-chroot
